eos-tech-support/eos-download-image

```diff
@@ -1,125 +1,114 @@
+    subprocess.check_call(args, stdout=sys.stderr)
+
+
+def ensure_keyring():
+    if os.path.isfile(KEYRING_SYSTEM_PATH):
+        return KEYRING_SYSTEM_PATH
+
+    keyring = os.path.abspath(KEYRING_FILENAME)
+    if not os.path.isfile(keyring):
+        log("Fetching Endless keyring")
+        check_call("wget", "-O", keyring, KEYRING_URL)
+
+    return keyring
+
+
+def download_and_verify(image_url, signature_url):
+    log("Downloading image from", image_url)
+    check_call("wget", "--continue", image_url)
+
+    log("Downloading signature from", signature_url)
+    check_call("wget", "--continue", signature_url)
+
+    keyring = ensure_keyring()
+
+    log("Verifying")
+    check_call("gpgv", "--keyring", keyring,
+               os.path.basename(signature_url),
+               os.path.basename(image_url))
+
+    log("Download okay.")
+    return os.path.basename(image_url)
+
+
+def main():
+    p = argparse.ArgumentParser(
+        description='Fetches and verifies an image (installer or regular OS)')
+    p.add_argument('-v', '--version',
+                   help='Image version (default: newest)')
+    p.add_argument('-p', '--personality',
+                   help='Image personality (default: base)',
+                   default='base')
+    p.add_argument('-r', '--product',
+                   help='Product (eg: eosinstaller; default: eos)',
+                   default='eos')
+    p.add_argument('-i', '--internal',
+                   action='store_true',
+                   help='Fetch images from the Endless internal network')
+
+    args = p.parse_args()
+    base = INTERNAL_BASE_URL if args.internal else PUBLIC_BASE_URL
+
+    manifest_url = '{base}/releases-{product}.json.gz'.format(
+        base=base, product=args.product)
+    req = urllib.request.Request(manifest_url)
+    with urllib.request.urlopen(req) as conn:
+        with gzip.GzipFile(mode='r', fileobj=conn) as manifest_fp:
+            manifest = json.load(codecs.getreader('utf8')(manifest_fp))
+
+    images = list(manifest['images'].values())
+    images.sort(key=lambda i: distutils.version.StrictVersion(i['version']))
+    if args.version is not None:
+        for selected in images:
+            if selected['version'] == args.version:
+                break
+        else:
+            log("Image version", repr(args.version), "not found")
+            log("Known versions:", *[i['version'] for i in images])
+            sys.exit(1)
+    else:
+        selected = images[-1]
+
+    personalities = selected['personality_images']
+    try:
+        personality = personalities[args.personality]
+    except KeyError:
+        log("Personality", repr(args.personality), "not found")
+        log("Known personalities:", *sorted(personalities.keys()))
+        sys.exit(1)
+
+    image_url = '{}/{}'.format(base, personality['full']['file'])
+    signature_url = '{}/{}'.format(base, personality['full']['signature'])
+    print(download_and_verify(image_url, signature_url))
```

Comment:
The `print` is to stdout: this script is invoked by `eos-write-installer` and `eos-write-live`, which read the image filename from this script's stdout.